torch/csrc/distributed/c10d/init.cpp

```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
           "wait",
           &::c10d::ProcessGroup::Work::wait,
           py::arg("timeout") = kNoTimeout,
-          py::call_guard<py::gil_scoped_release>());
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "get_future",
+          [](::c10d::ProcessGroup::Work& work)
+              -> std::shared_ptr<jit::PythonFutureWrapper> {
+            return std::make_shared<jit::PythonFutureWrapper>(work.getFuture());
+          },
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
+            ``get_future`` retrieves a future associated with the completion of
+            ``c10d.ProcessGroup.work``. As an example, a future object can be
+            retrieved by `fut = dist.allreduce(tensors).get_future()`.
+
+            In the example above, if `dist.allreduce` work was done on GPU,
+            `fut.wait()` would return after synchronizing the correct GPU streams
+            to ensure we can have async CUDA execution and it does not wait for
+            the entire operation to complete on GPU. If NCCL_BLOCKING_WAIT is enabled,
```

Comment:
> `fut.wait()` would return after synchronizing the correct GPU streams to ensure we can have async CUDA execution and it does not wait for the entire operation to complete on GPU.

what does it referring to here? And what does `synchronizing` mean here? I am assuming `synchronizing` does not mean blocking here? If so, please expand this to make that clear. Only saying `synchronizing` is a little confusing, as people might think it's similar to `torch.cuda.synchronize()`, which is blocking

Reply:
I think we wanna give a sense of what we did in NCCL here. My understanding in NCCL `synchronize` just refers to `cudaEvents_[i].block(currentStream);` cc @pritamdamania87 any thoughts?

Reply:
Yep, I actually understand what you mean here. Trying to review from a user's perspective, and assuming the user is not familiar `ProcessGroupNCCL` implementation.

Reply:
For users familiar with synchronization between streams on CUDA, shouldn't this be clear that we're synchronizing the correct streams here? We could expand this doc to mention that NCCL uses its own pool of streams and pytorch has its own default streams per device. Synchronizing here means that we're synchronizing the appropriate NCCL streams with the device streams such that subsequent operations on the return value are processed correctly by the default device stream.